### PR TITLE
SafariServices bindings for iOS 10 and macOS 10.12

### PR DESCRIPTION
**Pull request by timrisi**

### CODE REVIEW

**Comment by xamarin-release-manager** [comment]
Build success

File: src/safariservices.cs [hunk 1]
```diff
@@ -51,20 +79,29 @@ interface SFSafariViewController {
...
 		[Export ("initWithURL:entersReaderIfAvailable:")]
 		IntPtr Constructor (NSUrl url, bool entersReaderIfAvailable);
 
 		[Export ("initWithURL:")]
 		IntPtr Constructor (NSUrl url);
 
+		[iOS (10,0)]
+		[Export ("initWithURL:configuration:")]
+		[DesignatedInitializer]
+		IntPtr Constructor (NSUrl URL, SFSafariViewControllerConfiguration configuration);
```
Comment (spouliot) [inline src/safariservices.cs]: parameter name should be `url` in lower case

File: src/safariservices.cs [hunk 2]
```diff
@@ -81,4 +118,141 @@ partial interface SFSafariViewControllerDelegate {
...
+
+		[Export ("reload")]
+		void Reload ();
+
+		[Export ("getPagePropertiesWithCompletionHandler:")]
+		void GetPageProperties (Action<SFSafariPageProperties> completionHandler);
+	}
+
+	[Mac (10,12)]
+	[Protocol, Model]
```
Comment (spouliot) [inline src/safariservices.cs]: I'm not sure about how this is used... but is a `[Model]` needed ? or is this something that will likely be adopted by an `NSView` (or similar) ?

File: src/safariservices.cs [hunk 3]
```diff
@@ -81,4 +118,141 @@ partial interface SFSafariViewControllerDelegate {
...
+	interface SFSafariExtensionHandling
+	{
+		[Export ("messageReceivedWithName:fromPage:userInfo:")]
+		void MessageReceived (string messageName, SFSafariPage page, [NullAllowed] NSDictionary userInfo);
+
+		[Export ("toolbarItemClickedInWindow:")]
+		void ToolbarItemClicked (SFSafariWindow window);
+
+		[Export ("validateToolbarItemInWindow:validationHandler:")]
+		void ValidateToolbarItem (SFSafariWindow window, Action<bool, NSString> validationHandler);
```
Comment (spouliot) [inline src/safariservices.cs]: sounds like a case where `[Async]` would be useful

File: src/safariservices.cs [hunk 4]
```diff
@@ -81,4 +118,141 @@ partial interface SFSafariViewControllerDelegate {
...
+		bool Active { [Bind ("isActive")] get; }
+	}
+
+	[Mac (10,12)]
+	[BaseType (typeof(NSObject))]
+	[DisableDefaultCtor]
+	interface SFSafariTab
+	{
+		[Export ("getActivePageWithCompletionHandler:")]
+		void GetActivePage (Action<SFSafariPage> completionHandler);
```
Comment (spouliot) [inline src/safariservices.cs]: all 3 methods looks like `[Async]` candidate

File: src/safariservices.cs [hunk 5]
```diff
@@ -81,4 +118,141 @@ partial interface SFSafariViewControllerDelegate {
...
+		void SetEnabled (bool enabled, [NullAllowed] string badgeText);
+	}
+
+	[Mac (10,12)]
+	[BaseType (typeof(NSObject))]
+	[DisableDefaultCtor]
+	interface SFSafariWindow
+	{
+		[Export ("getActiveTabWithCompletionHandler:")]
+		void GetActiveTab (Action<SFSafariTab> completionHandler);
```
Comment (spouliot) [inline src/safariservices.cs]: same wrt `[Async]`

File: src/safariservices.cs [hunk 6]
```diff
@@ -81,4 +118,141 @@ partial interface SFSafariViewControllerDelegate {
...
+	interface SFSafariExtensionHandler : NSExtensionRequestHandling, SFSafariExtensionHandling
+	{
+	}
+
+	[Static]
+	partial interface SFContentBlocker
+	{
+		[Mac (10, 12)]
+		[Field ("SFContentBlockerErrorDomain")]
+		NSString ErrorDomain { get; }
```
Comment (spouliot) [inline src/safariservices.cs]: that should be located to the error enum using `[ErrorDomain]`

File: src/safariservices.cs [hunk 7]
```diff
@@ -51,20 +79,29 @@ interface SFSafariViewController {
 		[PostGet ("NibBundle")]
 		IntPtr Constructor ([NullAllowed] string nibName, [NullAllowed] NSBundle bundle);
 
-		[DesignatedInitializer]
+		[Deprecated (PlatformName.iOS, 10, 0, message: "Please use initWithURL:configuration:")]
 		[Export ("initWithURL:entersReaderIfAvailable:")]
```
Comment (migueldeicaza) [inline src/safariservices.cs]: This message would not make sense to a C# developer, replace with the proper constructor signature.

File: src/safariservices.cs [hunk 8]
```diff
@@ -81,4 +118,141 @@ partial interface SFSafariViewControllerDelegate {
...
+#else
+	[Mac (10,12)]
+	[BaseType (typeof(NSObject))]
+	[DisableDefaultCtor]
+	interface SFSafariApplication
+	{
+		[Static]
+		[Export ("getActiveWindowWithCompletionHandler:")]
+		void GetActiveWindow (Action<SFSafariWindow> completionHandler);
+
```
Comment (migueldeicaza) [inline src/safariservices.cs]: Add [Async] so that this produces an Async variant.

File: src/safariservices.cs [hunk 9]
```diff
@@ -81,4 +118,141 @@ partial interface SFSafariViewControllerDelegate {
...
+		[Static]
+		[Export ("openWindowWithURL:completionHandler:")]
+		void OpenWindow (NSUrl url, [NullAllowed] Action<SFSafariWindow> completionHandler);
+
```
Comment (migueldeicaza) [inline src/safariservices.cs]: Add [Async] attribute to get the async variant.

File: src/safariservices.cs [hunk 10]
```diff
@@ -81,4 +118,141 @@ partial interface SFSafariViewControllerDelegate {
...
+	{
+		[Export ("dispatchMessageToScriptWithName:userInfo:")]
+		void DispatchMessageToScript (string messageName, [NullAllowed] NSDictionary userInfo);
+
+		[Export ("reload")]
+		void Reload ();
+
+		[Export ("getPagePropertiesWithCompletionHandler:")]
+		void GetPageProperties (Action<SFSafariPageProperties> completionHandler);
+	}
```
Comment (migueldeicaza) [inline src/safariservices.cs]: Add [Async] for having an async variant.

**Comment by xamarin-release-manager** [comment]
Build success

**Pull request merged**